README/RCodingStyle.hxx

```diff
@@ -0,0 +1,178 @@
+/// \file RCodingStyle.hxx
+/// \author Axel Naumann <axel@cern.ch>
+/// \author Other Author <other@author>
+/// \date 2018-07-06
+// The above entries are mostly for giving some context.
+// The "author" field gives a hint whom to contact in case of questions, also
+// from within the team. The date shows whether this is ancient or only part
+// of the latest release. It's the date of creation / last massive rewrite.
+
+/*************************************************************************
+ * Copyright (C) 1995-2018, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+// This file demonstrates the coding styles to be used for new ROOT files.
+// See the accompanying source file for context.
+
+// File names correspond to the name of the main class that they contain.
+// Our users will expect that a file called "RCodingStyle.hxx" contains a class
+// named `RCodingStyle`.
+
+// All files start with a Doxygen header, followed by the copyright statement.
+
+// The file must contain no trailing whitespace including no indentation-only lines.
+// Except for extreme cases, line breaks should happen between 80 and 120 characters.
+
+// The first non-comment, non-empty line must be the #include guard.
+// Acceptable forms are:
+//   ROOT_RCodingStyle, ROOT_RCodingStyle_hxx, ROOT_RCodingStyle_Hxx,
+//   ROOT_RCODINGSTYLE, ROOT_RCODINGSTYLE_HXX
+// The first is preferred.
+#ifndef ROOT_RCodingStyle
+
+// Especially within header files, reduce the number of included headers as
+// much as possible, at the expense of more forward declarations.
+// EXCEPTION: standard library classes must be #included, not forward declared.
+
+// Include ROOT's headers first, in alphabetical order:
+// Rationale: make sure that ROOT headers are standalone.
+// Do not use #include guards at the point of `#include`.
+// Rationale: compilers optimize this themselves, this is interfering with their optimization.
+#include <ROOT/THist.hxx>
+
+// Include stdlib headers next, in alphabetical order.
+#include <string>
+
+// Include non-ROOT, non-stdlib headers last.
+// Rationale: non-ROOT, non-stdlib headers often #define like mad. Reduce interference
+// with ROOT or stdlib headers.
+#include <png.h>
```

Comment:
That sounds weird. In principle png.h may depend on stl. That's not a problem for non-ROOT packages but what about the ones that are on the borderline: eg libafterimage and others that are patched by us.

Reply:
> png.h may depend on stl

No. System headers never depend on C++ and should thus come last.